File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/LogSoftMax.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.transforms.custom;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.common.base.Preconditions;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;
import org.nd4j.linalg.api.ops.impl.transforms.gradient.LogSoftMaxDerivative;
import org.nd4j.linalg.factory.Nd4j;

import java.util.Collections;
import java.util.List;

/**
 * Log(softmax(X))
 *
 * @author Alex Black
 */

public class LogSoftMax extends DynamicCustomOp {

    private Integer dimension = null;

    public LogSoftMax(SameDiff sameDiff, SDVariable i_v) {
        super(sameDiff, i_v);
    }

    public LogSoftMax() {
    }

    public LogSoftMax(INDArray x, INDArray z) {
        super(null, x, z, null, null);
    }

    public LogSoftMax(INDArray x) {
        this(x, x);
    }

    public LogSoftMax(INDArray x, int dimension) {
        this(x, null);
        this.dimension = dimension;
    }

    public LogSoftMax(SameDiff sameDiff, SDVariable i_v, int dimension) {
        this(sameDiff, i_v);
        this.dimension = dimension;
        addIArgument(dimension);
    }


    @Override
    public String opName() {
        return "log_softmax";
    }
    @Override
    public String tensorflowName() {
        return "LogSoftmax";
    }

    @Override
    public List<SDVariable> doDiff(List<SDVariable> i_v) {
        if(dimension == null) {
            return new LogSoftMaxDerivative(sameDiff, arg(), i_v.get(0)).outputs();
        } else {
            return new LogSoftMaxDerivative(sameDiff, arg(), i_v.get(0), dimension).outputs();
        }
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> inTypes){
        Preconditions.checkState(inTypes != null && inTypes.size() == 1, "Expected 1 input datatype for %s, got %s",
                getClass(), inTypes);
        if(inTypes.get(0).isFPType())
            return Collections.singletonList(inTypes.get(0));
        return Collections.singletonList(Nd4j.defaultFloatingPointType());
    }
}

File: libnd4j/include/ops/declarable/generic/nn/logSoftmax.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// @author Yurii Shyrma (iuriish@yahoo.com), created on 01.02.2018
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_log_softmax)

#include <ops/declarable/CustomOperations.h>
#include <ops/declarable/helpers/activations.h>

namespace sd {
namespace ops {

DECLARE_TYPES(log_softmax) { getOpDescriptor()->setAllowedInputTypes({ALL_FLOATS})->setSameMode(true); }

CONFIGURABLE_OP_IMPL(log_softmax, 1, 1, true, 0, 0) {
  auto input = INPUT_VARIABLE(0);
  auto output = OUTPUT_VARIABLE(0);

  const int rank = input->rankOf();
  const int dim = block.getIArguments()->size() > 0 ? INT_ARG(0) : rank - 1;

  REQUIRE_TRUE(dim < rank, 0,
               "LOG_SOFTMAX OP: the value of input integer parameter (dimension) must be less than input array rank "
               "%i, but got dimension = %i instead !",
               rank, dim);

  helpers::logSoftmax(block.launchContext(), input, output, dim);

  return sd::Status::OK;
}

DECLARE_TYPES(log_softmax_bp) {
  getOpDescriptor()
      ->setAllowedInputTypes(0, DataType::ANY)
      ->setAllowedInputTypes(1, {ALL_FLOATS})
      ->setAllowedOutputTypes({ALL_FLOATS});
}

CONFIGURABLE_OP_IMPL(log_softmax_bp, 3, 1, true, 0, 0) {
  auto input = INPUT_VARIABLE(0);
  auto gradO = INPUT_VARIABLE(1);
  auto softmaxOut = INPUT_VARIABLE(2);
  auto gradI = OUTPUT_VARIABLE(0);
  gradI->assign(softmaxOut);
  const int rank = input->rankOf();
  const int dim = block.getIArguments()->size() > 0 ? INT_ARG(0) : rank - 1;

  REQUIRE_TRUE(dim < rank, 0,
               "LOG_SOFTMAX_BP OP: the value of input integer parameter (dimension) must be less than input array rank "
               "%i, but got dimension = %i instead !",
               rank, dim);

  helpers::softmax(block.launchContext(), input, gradI, dim);

  std::vector<sd::LongType> dimVec;
  dimVec.push_back(dim);
  auto sumGradOj = gradO->reduceAlongDimension(reduce::Sum,&dimVec, true);
  //we stored softmax inside gradI
  NDArray assign = *gradO - *gradI * sumGradOj;
  gradI->assign(&assign);

  return sd::Status::OK;
}

}  // namespace ops
}  // namespace sd

#endif
